File: T-Bank/Algorithms-and-Data-Structures-2025/10sprint-SegmentTree2/contest/F-Сад_пермского_периода/solution.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <vector>

struct Center {
    long long x;
    long long y;
    int id;
};

void solution() {
    int n;
    std::scanf("%*d %*d %d", &n);
    std::vector<Center> centers(n);
    for (int id = 0; id < n; ++id) {
        double x, y;
        std::scanf("%lf %lf", &x, &y);
        centers[id] = Center{(long long)(x * 2), (long long)(y * 2), id};
    }
    std::vector<long long> y(n);
    for (int i = 0; i < n; ++i) y[i] = centers[i].y;
    std::sort(y.begin(), y.end());
    y.erase(std::unique(y.begin(), y.end()), y.end());
    std::vector<long long> x(y.size(), 0);
    std::function<long long(int)> getX = [&](int index) {
        long long sum = 0;
        for (; index >= 0; index = (index & (index + 1)) - 1) {
            sum += x[index];
        }
        return sum;
    };
    std::function<void(int, long long)> addSinceX = [&](int index, long long extra) {
        for (; index < (int)x.size(); index = index | (index + 1)) {
            x[index] += extra;
        }
    };
    std::function<void(int, int, long long)> addRangeX = [&](int first, int after, long long extra) {
        addSinceX(first, extra);
        addSinceX(after, -extra);
    };
    std::sort(centers.begin(), centers.end(), [](const Center &left, const Center &right) {
        return left.x < right.x;
    });
    std::vector<long long> ans(n, 0);
    for (Center center : centers) {
        auto it = std::lower_bound(y.begin(), y.end(), center.y);
        int index = int(it - y.begin());
        long long dist = center.x - getX(index);
        ans[center.id] = dist;
        it = std::upper_bound(y.begin(), y.end(), center.y - dist);
        int first = int(it - y.begin());
        it = std::upper_bound(y.begin(), y.end(), center.y + dist);
        int after = int(it - y.begin());
        addRangeX(first, after, dist * 2);
    }
    for (long long value : ans) std::printf("%lld ", value);
}

int main() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    solution();
    
}

File: T-Bank/Algorithms-and-Data-Structures-2025/10sprint-SegmentTree2/contest/F-Сад_пермского_периода/solution.py
import sys
from typing import List, Callable
from dataclasses import dataclass


@dataclass
class Center:
    x: int
    y: int
    id: int


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n = int(sys.stdin.readline().rstrip().split()[2])
    centers: list[Center] = []
    for id in range(n):
        x, y = map(float, sys.stdin.readline().split())
        centers.append(Center(int(x * 2), int(y * 2), id))
    y_coords: list[int] = sorted(set(center.y for center in centers))
    x: list[int] = [0] * len(y_coords)
    
    def get_x(index: int) -> int:
        sum_val: int = 0
        while index >= 0:
            sum_val += x[index]
            index = (index & (index + 1)) - 1
        return sum_val
    
    def add_since_x(index: int, extra: int) -> None:
        while index < len(x):
            x[index] += extra
            index = index | (index + 1)
    
    def add_range_x(first: int, after: int, extra: int) -> None:
        add_since_x(first, extra)
        add_since_x(after, -extra)
    
    centers.sort(key=lambda center: center.x)
    ans: list[int] = [0] * n
    for center in centers:
        index: int = bisect_left(y_coords, center.y)
        dist: int = center.x - get_x(index)
        ans[center.id] = dist
        first: int = bisect_right(y_coords, center.y - dist)
        after: int = bisect_right(y_coords, center.y + dist)
        add_range_x(first, after, dist * 2)
    sys.stdout.write(' '.join(map(str, ans)) + '\n')


def bisect_left(a: list[int], x: int) -> int:
    lo, hi = 0, len(a)
    while lo < hi:
        mid: int = lo + ((hi - lo) >> 1)
        if a[mid] < x: lo = mid + 1
        else: hi = mid
    return lo


def bisect_right(a: list[int], x: int) -> int:
    lo, hi = 0, len(a)
    while lo < hi:
        mid: int = lo + ((hi - lo) >> 1)
        if x < a[mid]: hi = mid
        else: lo = mid + 1
    return lo


if __name__ == "__main__":
    solution()
